validator/client/wait_for_activation.go

```diff
@@ -40,99 +41,132 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
+	// "-1" indicates that validator count endpoint is not supported by the beacon node.
+	var valCount int64 = -1
+	valCounts, err := v.prysmChainClient.ValidatorCount(ctx, "head", []validator2.Status{validator2.Active})
+	if err != nil && !errors.Is(err, iface.ErrNotSupported) {
+		return -1, errors.Wrap(err, "could not get active validator count")
+	}
+	if len(valCounts) > 0 {
+		valCount = int64(valCounts[0].Count)
+	}
+	return valCount, nil
+}
+
+func (v *validator) handleReconnection(ctx context.Context, span octrace.Span, err error, message string, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
+	tracing.AnnotateError(span, err)
+	attempts := streamAttempts(ctx)
+	log.WithError(err).WithField("attempts", attempts).Error(message)
+	// Reconnection attempt backoff, up to 60s.
+	time.Sleep(time.Second * time.Duration(math.Min(uint64(attempts), 60)))
+	return v.internalWaitForActivation(incrementRetries(ctx), accountsChangedChan)
+}
 
-	someAreActive := false
-	for !someAreActive {
-		select {
-		case <-ctx.Done():
-			log.Debug("Context closed, exiting fetching validating keys")
-			return ctx.Err()
-		case <-accountsChangedChan:
-			// Accounts (keys) changed, restart the process.
-			return v.internalWaitForActivation(ctx, accountsChangedChan)
-		default:
-			res, err := (stream).Recv() // retrieve from stream one loop at a time
-			// If the stream is closed, we stop the loop.
-			if errors.Is(err, io.EOF) {
-				break
-			}
-			// If context is canceled we return from the function.
-			if errors.Is(ctx.Err(), context.Canceled) {
-				return errors.Wrap(ctx.Err(), "context has been canceled so shutting down the loop")
-			}
-			if err != nil {
-				tracing.AnnotateError(span, err)
-				attempts := streamAttempts(ctx)
-				log.WithError(err).WithField("attempts", attempts).
-					Error("Stream broken while waiting for activation. Reconnecting...")
-				// Reconnection attempt backoff, up to 60s.
-				time.Sleep(time.Second * time.Duration(math.Min(uint64(attempts), 60)))
-				return v.internalWaitForActivation(incrementRetries(ctx), accountsChangedChan)
-			}
-
-			for _, s := range res.Statuses {
-				v.pubkeyToStatus[bytesutil.ToBytes48(s.PublicKey)] = &validatorStatus{
-					publicKey: s.PublicKey,
-					status:    s.Status,
-					index:     s.Index,
-				}
-			}
-
-			// "-1" indicates that validator count endpoint is not supported by the beacon node.
-			var valCount int64 = -1
-			valCounts, err := v.prysmChainClient.ValidatorCount(ctx, "head", []validator2.Status{validator2.Active})
-			if err != nil && !errors.Is(err, iface.ErrNotSupported) {
-				return errors.Wrap(err, "could not get active validator count")
-			}
-
-			if len(valCounts) > 0 {
-				valCount = int64(valCounts[0].Count)
-			}
-
-			someAreActive = v.checkAndLogValidatorStatus(valCount)
+func (v *validator) waitForAccountsChange(ctx context.Context, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
```

Comment:
We have some of these unexported functions with comments and others not. Should we just add comments to the rest of them?

Reply:
i don't think comments are required on unexported, it's probably fine to keep whatever we have.